File: js/src/model/conditions/SimplexRule.js
import RuleValidateResult from '../matching/RuleValidateResult'
import { R } from '../enums'
import { RuleRange } from '../Condition'

/**
 * 单一条件规则, [#cPID1#cPID2].sum(4000)
 */
class SimplexRule {
    range;
    predict;
    expected;
    static calculatorPattern = /^(?<range>[\[\]#\w:\-\u4e00-\u9fa5]+|[\$~])\.(?<predict>\w+)_(?<threshold>\d+)$/;
    static rulePattern = /^([|&(]?(\[(#[cpk]\w+|#z[^#\]]+)+\]|[\$~])\.\w+\(\d+\)\)?)+$/;

    /**
     * 单一条件规则, [#cPID1#cPID2].sum(4000)
     * @param {RuleRangeCollection} range
     * @param {P} predict
     * @param {int} expected
     */
    constructor(range, predict, expected){
        this.range = range;
        this.predict = predict;
        this.expected = expected;
    }

    /**
     * 字符串格式化
     * @returns {string}
     */
    toRuleString() {
        return `${this.range.toString()}.${this.predict.toString()}(${this.expected})`;
    }

    /**
     * 字符串格式化
     * @returns {string}
     */
    toString() {
        return this.toRuleString();
    }

    /**
     * 检查items是否满足当前规则
     * @param {Item []} items
     * @returns {boolean}
     */
    check(items) {
        return this._getActual(items)>=this.expected;
    }

    /**
     * 检查items是否满足当前规则,并且返回详细的核对结果
     * @param {Item []} items
     * @returns {RuleValidateResult}
     */
    validate(items) {
        let actual = this._getActual(items);
        return new RuleValidateResult(actual>=this.expected,
            this,this.expected,actual);
    }

    _getActual(items){
        items =  this.filterItem (items);
        return this.predict.handler(items);
    }

    /**
     * 根据range过滤出当前规则所匹配的item
     * @param {Item []} items
     * @returns {Item []}
     */
    filterItem (items){
        if(this.range == R.ALL){
            return items;
        }
        else{
            let wheres = [];
            for(const match of this.range.toString().matchAll(RuleRange.rangePattern)){
                let cate = match.groups["cate"];
                let item = match.groups["item"];
                switch(cate){
                    case 'c':
                        let f1 = function(cate){
                            return function(item){
                                return item.category === cate;
                            }
                        }(item);
                        wheres.push(f1);
                        break;
                    case 'p':
                        let f2 = function(spu){
                            return function(item){
                                return item.SPU === spu;
                            }
                        }(item);
                        wheres.push(f2);
                        break;
                    case 'k':
                        let f3 = function(sku){
                            return function(item){
                                return item.SKU === sku;
                            }
                        }(item);
                        wheres.push(f3);
                        break;
                    case 'z':
                        let f4 = function(seatRange){
                            return function(item){
                                return new RuleRange('z',seatRange).isSeatInRange(item);
                            }
                        }(item)
                        wheres.push(f4);
                        break;
                    default:
                        throw "不支持的规则判断范围";
                }
            }
            return items.filter(t=>{
                for(let f of wheres){
                    if(f(t)){
                        return true;
                    }
                }
                return false;
            });
        }
    }
}

export default SimplexRule;

File: java/src/main/java/com/github/thinhunan/wonder8/promotion/rule/model/SimplexRule.java
package com.github.thinhunan.wonder8.promotion.rule.model;

import com.github.thinhunan.wonder8.promotion.rule.model.validate.RuleValidateResult;

import java.util.List;
import java.util.function.Predicate;

/**
 * 单一条件规则, [#cPID1#cPID2].sum(4000)
 */
public class SimplexRule extends RuleComponent {
    //判断范围
    RuleRangeCollection range;
    //判断方式
    P predict;
    //判断标准
    int expected;

    public RuleRangeCollection getRange() {
        return range;
    }

    public void setRange(RuleRangeCollection range) {
        this.range = range;
    }

    public P getPredict() {
        return predict;
    }

    public void setPredict(P predict) {
        this.predict = predict;
    }

    public int getExpected() {
        return expected;
    }

    public void setExpected(int expected) {
        this.expected = expected;
    }

    boolean _selfCheck(){
        return !(this.range == null || this.range.size() < 1 || expected < 1);
    }

    @Override
    public boolean check(List<Item> items) {
        if(!_selfCheck()){
            return false;
        }
        return _getActual(items) >= expected;
    }

    @Override
    public RuleValidateResult validate(List<Item> items) {
        if(!_selfCheck()){
            return new RuleValidateResult(false,null,0,0,null );
        }
        int actual = _getActual(items);
        return RuleValidateResult.builder()
                .rule(this)
                .actual(actual)
                .expected(expected)
                .valid(actual >= expected)
                .build();
    }

    private int _getActual(List<Item> items) {

        Predicate<Item> fitted = getFilter();
        return Validator.getValidator(this.getPredict()).apply(items.stream().filter(fitted));
    }

    /**
     * 根据range过滤当前规则所匹配的那部分ticket
     */
     public Predicate<Item> getFilter() {
        Predicate<Item> filter = t->false;
        for(RuleRange r : this.range) {
            switch (r.getType()) {
                case CATEGORY:
                    filter = filter.or(t -> t.getCategory().equals(r.getId()));
                    break;
                case SPU:
                    filter = filter.or(t -> t.getSPU().equals(r.getId()));
                    break;
                case SKU:
                    filter = filter.or(t -> t.getSKU().equals(r.getId()));
                    break;
                case SEAT:
                    filter = filter.or(t-> r.isSeatInRange(t));
                    break;
                default:
                    filter = t -> true;
                    break;
            }
        }
        return filter;
    }


    @Override
    public String toRuleString() {
        return String.format("%s.%s(%d)",range.toRuleString(), predict,expected);
    }

    @Override
    public String toString(){
        return this.toRuleString();
    }
}
